library/net.cpp

```diff
@@ -1424,3 +1433,275 @@ int dnet_trans_forward(struct dnet_io_req *r, struct dnet_net_state *orig, struc
+		              dnet_addr_string_raw(&forward->addr, daddr, sizeof(daddr)),
+		              t->rcv_trans, t->trans);
+	}
+
+	forward->n2_tmp_forwarding_in_progress = 1;
+
+	return n2_trans_send(t, request_info);
+}
+
+int n2_send_error_response(struct dnet_net_state *st,
+                           struct n2_request_info *req_info,
+                           int errc,
+                           struct dnet_access_context *context) {
+	auto impl = [&] {
+		req_info->repliers.on_reply_error(errc);
+		return 0;
+	};
+	return c_exception_guard(impl, st->n, __FUNCTION__);
+}
+
+class cork_guard_t {
+public:
+	explicit cork_guard_t(int write_socket)
+	: write_socket_(write_socket)
+	{
+		ser_cork(1);
+	}
+
+	~cork_guard_t() {
+		ser_cork(0);
+	}
+
+private:
+	void ser_cork(int cork) {
+		setsockopt(write_socket_, IPPROTO_TCP, TCP_CORK, &cork, 4);
+	}
+
+	const int write_socket_;
+};
+
+int n2_send_request_impl(struct dnet_net_state *st, struct dnet_io_req *r) {
+	using namespace ioremap::elliptics;
+
+	if (st->send_offset == 0) {
+		clock_gettime(CLOCK_MONOTONIC_RAW, &st->send_start_ts);
+		r->queue_time = DIFF_TIMESPEC(r->queue_start_ts, st->send_start_ts);
+	}
+
+	int send_error = 0;
+	uint64_t send_time = 0;
+	const n2::net_iovec &iov = r->iov->iov;
+
+	// In net_iovec dnet_cmd is represented in net bytes order. So here we convert its copy to cpu bytes order,
+	// to have access to some dnet_cmd fields
+	dnet_cmd cmd = *iov.front().data<dnet_cmd>();
+	dnet_convert_cmd(&cmd);
+	
+	uint64_t total_size = sizeof(dnet_cmd) + cmd.size;
+
+	dnet_logger_set_trace_id(cmd.trace_id, cmd.flags & DNET_FLAGS_TRACE_BIT);
+	const enum dnet_log_level level = st->send_offset == 0 ? DNET_LOG_NOTICE : DNET_LOG_DEBUG;
+	DNET_LOG(st->n, level, "%s: %s: sending trans: %lld -> %s/%d: size: %llu, cflags: %s, start-sent: "
+			       "%zd/%zd, send-queue-time: %lu usecs",
+		 dnet_dump_id(&cmd.id), dnet_cmd_string(cmd.cmd), (unsigned long long)cmd.trans,
+		 dnet_addr_string(&st->addr), cmd.backend_id, (unsigned long long)cmd.size,
+		 dnet_flags_dump_cflags(cmd.flags), st->send_offset, total_size, r->queue_time);
+
+	{
+		cork_guard_t cork_guard(st->write_s);
+
+		size_t current_block_offset = 0;
+
+		for (const auto &dp : iov) {
+			size_t current_block_end_offset = current_block_offset + dp.size();
+			if (st->send_offset < current_block_end_offset) /*block hasn't sent*/ {
+				auto dp_left = dp.skip(st->send_offset - current_block_offset);
+				send_error = dnet_send_nolock(st, dp_left.data(), dp_left.size());
+				if (send_error)
+					break;
+			}
+
+			current_block_offset = current_block_end_offset;
+		}
+
+		// Logging block
+		struct timespec ts;
+		clock_gettime(CLOCK_MONOTONIC_RAW, &ts);
+		send_time = DIFF_TIMESPEC(st->send_start_ts, ts);
+		enum dnet_log_level level = DNET_LOG_DEBUG;
+		if (!send_error) {
+			level = !(cmd.flags & DNET_FLAGS_MORE) ? DNET_LOG_INFO : DNET_LOG_NOTICE;
+			dnet_access_context_add_uint(r->context, "send_time", send_time);
```

Comment:
`dnet_access_context_add_uint` is C function, here `r->context->add(...)` could be used instead.

Reply:
Done.